packages/library/src/commonMain/kotlin/io/realm/internal/PropertyDescriptor.kt

```diff
@@ -0,0 +1,170 @@
+
+// Realm field type represents the type of a given user specified field in the RealmObject
+data class RealmFieldType(
+    val collectionType: CollectionType,
+    val elementType: ElementType
+) {
+    val isPrimaryKeySupported: Boolean =
+        collectionType == CollectionType.RLM_COLLECTION_TYPE_NONE && elementType.realmFieldType.primaryKeySupport
+
+    // Utility method to generate Kotlin code for the specific field
+    fun toKotlinLiteral(): String {
+        val elementType = this.elementType
+        val element = (elementType.classifier as KClass<*>).simpleName + (if (elementType.nullable) "?" else "")
+        return when (collectionType) {
+            CollectionType.RLM_COLLECTION_TYPE_NONE -> element
+            CollectionType.RLM_COLLECTION_TYPE_LIST -> "List<$element>"
+            CollectionType.RLM_COLLECTION_TYPE_SET -> TODO()
+            CollectionType.RLM_COLLECTION_TYPE_DICTIONARY -> TODO()
+        }
+    }
+
+    override fun toString(): String {
+        return "RType(collectionType=$collectionType, elementType=$elementType)"
+    }
+}
+
+// Convenience methods to easily derive Realm field information from Kotlin types.
+fun KType.rType(): RealmFieldType {
+    val elementType = elementType(this)
+    return RealmFieldType(
+        collectionType(this),
+        ElementType(elementType.classifier!!, elementType.isMarkedNullable)
+    )
+}
+fun KMutableProperty1<*, *>.rType(): RealmFieldType {
+    return this.returnType.rType()
+}
+
+// Convenience class to easily derive information about a Realm field directly from the property.
+// It is unclear if we can derive sufficient information without access to annotations at runtime,
+// but alternatively we can maybe query information from the schema and key cache infrastructure.
+class RealmFieldDescriptor(val property: KMutableProperty1<*, *>) {
+    val rType by lazy { property.rType() }
+
+    val isElementNullable: Boolean = rType.elementType.nullable
+
+    // TODO Annotations are not available at runtime on Kotlin native
+    // val isPrimariKey: Boolean =
+    //    rType.isPrimaryKeySupported && property.annotations.isNotEmpty() && property.annotations[0] is PrimaryKey
+
+    // TODO Public/internal name. We cannot pull the public name for when obfuscated
+}
+
+private fun collectionType(type: KType): CollectionType {
+    return when (type.classifier) {
+        Set::class -> CollectionType.RLM_COLLECTION_TYPE_SET
+        List::class -> CollectionType.RLM_COLLECTION_TYPE_LIST
+        Map::class -> CollectionType.RLM_COLLECTION_TYPE_DICTIONARY
+        else -> CollectionType.RLM_COLLECTION_TYPE_NONE
+    }
+}
+
+private fun elementType(type: KType) = when (collectionType(type)) {
+    CollectionType.RLM_COLLECTION_TYPE_NONE ->
+        type
+    CollectionType.RLM_COLLECTION_TYPE_SET,
+    CollectionType.RLM_COLLECTION_TYPE_LIST ->
+        type.arguments[0].type!!
+    CollectionType.RLM_COLLECTION_TYPE_DICTIONARY ->
+        type.arguments[1].type!!
```

Comment:
Does this mean we'll be supporting String only as keys?

Reply:
That is at least the current support level in core and hence also Realm Java. I think it arises from sync where the keys are just the JSON keys. I guess we can just adjust our descriptors when the restriction is lifted.